File: services/visual-rag-client/QdrantAdapter.js
/**
 * QdrantAdapter.js
 *
 * Singleton adapter for Qdrant vector database operations.
 * Enforces Distance Metric Locks and integrates with global Circuit Breaker.
 *
 * Architecture Reference: Native Protocol Alpha-9
 * Epic: Structured Feature Development Workflow
 * Ticket: P1.1 Implement Singleton QdrantAdapter with Distance Metric Locks
 */

const { QdrantClient } = require('@qdrant/js-client-rest');
const { CircuitBreaker } = require('../experts/CircuitBreaker');
const logger = require('../logger');
const config = require('../../config/config');

const COLLECTIONS = {
    document_embeddings: { name: 'document_embeddings', vectorSize: 384, distance: 'Cosine' },
    visual_overlays: { name: 'visual_overlays', vectorSize: 320, distance: 'Cosine' },
    visual_pages: { name: 'visual_pages', vectorSize: 320, distance: 'Dot' }
};

class QdrantAdapter {
    constructor(options = {}) {
        if (QdrantAdapter.instance && Object.keys(options).length === 0) {
            return QdrantAdapter.instance;
        }

        // Initialize configuration
        const qdrantConfig = config.qdrant || {};
        const host = options.host || process.env.QDRANT_HOST || qdrantConfig.host || 'qdrant';
        const port = options.port || parseInt(process.env.QDRANT_PORT || qdrantConfig.port || '6333');
        const apiKey = options.apiKey || process.env.QDRANT_API_KEY || qdrantConfig.apiKey;
        const url = `http://${host}:${port}`;

        // Initialize Qdrant Client
        const checkCompatibility = (qdrantConfig && typeof qdrantConfig.checkCompatibility !== 'undefined') ? qdrantConfig.checkCompatibility : (process.env.QDRANT_CHECK_COMPATIBILITY !== 'false');
        this.client = new QdrantClient({
            url,
            apiKey,
            checkCompatibility
        });

        let clientVersion = 'unknown';
        try {
            clientVersion = require('@qdrant/js-client-rest/package.json').version;
        } catch (err) {
            logger.warn('[QdrantAdapter] Could not read @qdrant/js-client-rest version: ' + (err && err.message));
        }

        logger.info(`[QdrantAdapter] Initialized client at ${url} (client=${clientVersion}, checkCompatibility=${checkCompatibility})`);

        // Initialize Circuit Breaker for Qdrant operations
        // We use a dedicated breaker instance for vector DB operations
        this.circuitBreaker = CircuitBreaker.getInstance('qdrant-adapter', {
            failureThreshold: 5,
            cooldownPeriod: 30000,
            timeout: 10000, // Vector operations can be heavy
        });

        logger.info(`[QdrantAdapter] Initialized client at ${url}`);

        if (Object.keys(options).length === 0) {
            QdrantAdapter.instance = this;
        }
    }

    /**
     * Get the singleton instance
     * @returns {QdrantAdapter}
     */
    static getInstance() {
        if (!QdrantAdapter.instance) {
            QdrantAdapter.instance = new QdrantAdapter();
        }
        return QdrantAdapter.instance;
    }

    /**
     * Ensure a collection exists with the correct configuration (Distance Metric Lock)
     * @param {string} name - Collection name
     * @param {Object} vectorConfig - Vector configuration (size, distance)
     * @returns {Promise<void>}
     */
    async ensureCollection(name, vectorConfig) {
        const result = await this.circuitBreaker.execute(async () => {
            const exists = await this._collectionExists(name);

            if (exists) {
                const info = await this.client.getCollection(name);
                this._validateDistanceMetric(name, info, vectorConfig);
                logger.info(`[QdrantAdapter] Collection '${name}' verified`);
            } else {
                logger.info(`[QdrantAdapter] Creating collection '${name}'`);
                await this.client.createCollection(name, {
                    vectors: vectorConfig
                });
            }
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Failed to ensure collection ${name}`);
        }
    }

    /**
     * Batch upsert points
     * @param {string} collection 
     * @param {Array} points 
     * @returns {Promise<Object>}
     */
    async upsert(collection, points) {
        const result = await this.circuitBreaker.execute(async () => {
            return await this.client.upsert(collection, {
                points
            });
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Upsert failed for ${collection}`);
        }

        // Normalize to legacy helper response: { status: 'ok', count: N }
        const count = Array.isArray(points) ? points.length : (points && points.points ? points.points.length : 1);
        return { status: 'ok', count };
    }

    /**
     * Search for points
     * @param {string} collection 
     * @param {Object} options - { vector, filter, limit, with_payload }
     * @returns {Promise<Array>}
     */
    async search(collection, options) {
        const result = await this.circuitBreaker.execute(async () => {
            return await this.client.search(collection, options);
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Search failed for ${collection}`);
        }
        return result.data;
    }

    /**
     * Delete points by document ID
     * @param {string} collection 
     * @param {number|string} docId 
     * @returns {Promise<Object>}
     */
    async deleteByDocId(collection, docId) {
        const result = await this.circuitBreaker.execute(async () => {
            return await this.client.delete(collection, {
                filter: {
                    must: [
                        {
                            key: 'doc_id',
                            match: {
                                value: docId
                            }
                        }
                    ]
                }
            });
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Delete failed for ${collection} docId=${docId}`);
        }

        // Return legacy-style success object
        return { status: 'ok' };
    }

    /**
     * Delete points by explicit point IDs
     * @param {string} collection
     * @param {Array<string|number>} pointIds
     * @returns {Promise<Object>}
     */
    async deleteByIds(collection, pointIds) {
        const ids = Array.isArray(pointIds) ? pointIds : [pointIds];
        const result = await this.circuitBreaker.execute(async () => {
            return await this.client.delete(collection, {
                points: ids
            });
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Delete failed for ${collection} ids=${ids.length}`);
        }

        return { status: 'ok' };
    }

    /**
     * Update payload for a document across standard collections
     * @param {number|string} docId 
     * @param {Object} metadata 
     * @returns {Promise<Object>}
     */
    async updatePayload(docId, metadata) {
        const collections = ['visual_pages', 'visual_overlays'];
        const errors = [];
        const results = {};

        // Fail fast if circuit is open
        if (this.circuitBreaker.isOpen()) {
            throw new Error('Circuit breaker is OPEN, skipping payload update');
        }

        for (const collection of collections) {
            try {
                // 1. Find points for this doc_id
                const points = await this._getPointsByDocId(collection, docId);
                
                if (points.length === 0) {
                    results[collection] = { status: 'skipped', reason: 'no_points_found' };
                    continue;
                }

                const pointIds = points.map(p => p.id);

                // 2. Update payload
                const result = await this.circuitBreaker.execute(async () => {
                    await this.client.setPayload(collection, {
                        points: pointIds,
                        payload: metadata
                    });
                });

                if (result.fallback || !result.success) {
                    throw result.error || new Error(`Payload update failed for ${collection}`);
                }

                results[collection] = { status: 'updated', count: pointIds.length };

            } catch (err) {
                logger.error(`[QdrantAdapter] Failed to update payload for ${collection} docId=${docId}: ${err.message}`);
                errors.push({ collection, error: err.message });
                results[collection] = { status: 'failed', error: err.message };
            }
        }

        if (errors.length > 0) {
            const errorMsg = `Partial sync failure: ${errors.map(e => `${e.collection} (${e.error})`).join(', ')}`;
            const error = new Error(errorMsg);
            error.results = results;
            throw error;
        }

        return results;
    }

    /**
     * Get all collections
     * @returns {Promise<Object>}
     */
    async getCollections() {
        const result = await this.circuitBreaker.execute(async () => {
            return await this.client.getCollections();
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error('Failed to list collections');
        }
        return result.data;
    }

    /**
     * Get specific collection info
     * @param {string} name 
     * @returns {Promise<Object>}
     */
    async getCollection(name) {
        const result = await this.circuitBreaker.execute(async () => {
            return await this.client.getCollection(name);
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Failed to get collection info for ${name}`);
        }
        return result.data;
    }

    // =========================================================================
    // Private Helpers
    // =========================================================================

    async _collectionExists(name) {
        try {
            const result = await this.client.getCollections();
            return result.collections.some(c => c.name === name);
        } catch (error) {
            logger.error(`[QdrantAdapter] Failed to list collections: ${error.message}`);
            throw error;
        }
    }

    async _getPointsByDocId(collection, docId) {
        const result = await this.circuitBreaker.execute(async () => {
            let allPoints = [];
            let offset = null;
            
            do {
                const response = await this.client.scroll(collection, {
                    filter: {
                        must: [{ key: 'doc_id', match: { value: docId } }]
                    },
                    limit: 100,
                    offset: offset,
                    with_payload: false,
                    with_vector: false
                });
                
                allPoints = allPoints.concat(response.points);
                offset = response.next_page_offset;
                
            } while (offset);
            
            return allPoints;
        });

        if (result.fallback || !result.success) {
            throw result.error || new Error(`Failed to scroll points for ${collection}`);
        }
        return result.data;
    }

    _validateDistanceMetric(name, info, expectedConfig) {
        const actualVectors = info.config.params.vectors;
        
        // Helper to check distance match
        const checkDistance = (actual, expected) => {
            if (actual && expected && actual.toLowerCase() !== expected.toLowerCase()) {
                throw new Error(`Distance Metric Lock violation: Collection '${name}' has distance '${actual}' but expected '${expected}'. This is a critical error - MaxSim scoring requires correct distance metric.`);
            }
        };

        // Case 1: Expected is simple configuration (has distance property directly)
        if (expectedConfig.distance) {
            // If actual is named vectors (map) but we expect simple, it's a mismatch
            if (actualVectors && !actualVectors.distance && Object.keys(actualVectors).length > 0) {
                throw new Error(`Distance Metric Lock violation: Collection '${name}' has named vectors but expected simple configuration.`);
            }

            // Ensure actual has a distance property when expecting a simple config
            if (!actualVectors || !actualVectors.distance) {
                throw new Error(`Distance Metric Lock violation: Collection '${name}' does not expose a simple vector distance property.`);
            }

            checkDistance(actualVectors.distance, expectedConfig.distance);
            return;
        }

        // Case 2: Expected is named vectors (object with keys)
        const expectedKeys = Object.keys(expectedConfig);
        for (const key of expectedKeys) {
            const actualKey = actualVectors ? actualVectors[key] : null;
            if (!actualKey) {
                throw new Error(`Distance Metric Lock violation: Collection '${name}' missing expected vector '${key}'.`);
            }
            if (!actualKey.distance) {
                throw new Error(`Distance Metric Lock violation: Collection '${name}' vector '${key}' missing distance property.`);
            }
            checkDistance(actualKey.distance, expectedConfig[key].distance);
        }
    }

    // =========================================================================
    // Legacy / Helper Methods (Restored for Compatibility)
    // =========================================================================

    async initialize() {
        await this.ensureCollection(COLLECTIONS.document_embeddings.name, {
            size: COLLECTIONS.document_embeddings.vectorSize,
            distance: COLLECTIONS.document_embeddings.distance
        });
        await this.ensureCollection(COLLECTIONS.visual_overlays.name, {
            size: COLLECTIONS.visual_overlays.vectorSize,
            distance: COLLECTIONS.visual_overlays.distance
        });
        await this.ensureCollection(COLLECTIONS.visual_pages.name, {
            page_embedding: {
                size: COLLECTIONS.visual_pages.vectorSize,
                distance: COLLECTIONS.visual_pages.distance
            }
        });
    }

    async healthCheck() {
        try {
            const collections = await this.getCollections();
            const collectionNames = (collections && collections.collections) ? collections.collections.map(c => c.name) : [];
            
            const details = {};
            for (const col of Object.values(COLLECTIONS)) {
                if (collectionNames.includes(col.name)) {
                    // Presence in getCollections is sufficient to mark 'exists'
                    details[col.name] = { exists: true };
                    // Try to fetch extra details if available, but tolerate failures
                    try {
                        if (typeof this.getCollection === 'function') {
                            const info = await this.getCollection(col.name);
                            const vectors = info?.config?.params?.vectors || {};
                            const size = vectors.size || (vectors.page_embedding ? vectors.page_embedding.size : 'unknown');
                            const distance = vectors.distance || (vectors.page_embedding ? vectors.page_embedding.distance : 'unknown');

                            details[col.name] = {
                                exists: true,
                                pointCount: info.points_count,
                                vectorSize: size,
                                distance: distance
                            };
                        }
                    } catch (innerErr) {
                        // Keep exists=true and move on
                        logger.warn(`[QdrantAdapter] Could not fetch collection details for ${col.name}: ${innerErr.message}`);
                    }
                } else {
                    details[col.name] = { exists: false };
                }
            }

            return { healthy: true, collections: details };
        } catch (error) {
            return { healthy: false, error: error.message };
        }
    }

    // Visual Overlays Helpers
    async upsertVisualOverlays(points) {
        const normalizedPoints = (points || []).map(point => {
            if (point.vector) {
                return point;
            }
            if (!point.embedding) {
                return point;
            }
            const { embedding, ...rest } = point;
            return {
                ...rest,
                vector: embedding
            };
        });
        return this.upsert(COLLECTIONS.visual_overlays.name, normalizedPoints);
    }

    async searchVisualOverlays(vector, options = {}) {
        return this.search(COLLECTIONS.visual_overlays.name, {
            vector,
            limit: options.limit || 10,
            with_payload: true,
            score_threshold: options.scoreThreshold
        });
    }

    async deleteVisualOverlaysByDocId(docId) {
        return this.deleteByDocId(COLLECTIONS.visual_overlays.name, docId);
    }

    // Visual Pages Helpers
    async upsertVisualPages(points) {
        const normalizedPoints = (points || []).map(point => {
            if (point.vector) {
                return point;
            }
            if (!point.embedding) {
                return point;
            }
            const { embedding, ...rest } = point;
            return {
                ...rest,
                vector: {
                    page_embedding: embedding
                }
            };
        });
        return this.upsert(COLLECTIONS.visual_pages.name, normalizedPoints);
    }

    async searchVisualPages(vector, options = {}) {
        const candidates = [
            {
                vector: {
                    page_embedding: vector
                }
            },
            {
                vector,
                vector_name: 'page_embedding'
            },
            {
                vector: {
                    name: 'page_embedding',
                    vector
                }
            }
        ];

        for (const payload of candidates) {
            try {
                return await this.search(COLLECTIONS.visual_pages.name, {
                    ...payload,
                    limit: options.limit || 10,
                    with_payload: true
                });
            } catch (error) {
                if (!error.message || !error.message.includes('Bad Request')) {
                    throw error;
                }
            }
        }

        throw new Error('Bad Request');
    }

    async deleteVisualPagesByDocId(docId) {
        return this.deleteByDocId(COLLECTIONS.visual_pages.name, docId);
    }

    // Document Embeddings Helpers
    async upsertDocumentEmbeddings(points) {
        const normalizedPoints = (points || []).map(point => {
            if (point.vector) {
                return point;
            }
            if (!point.embedding) {
                return point;
            }
            const { embedding, ...rest } = point;
            return {
                ...rest,
                vector: embedding
            };
        });
        return this.upsert(COLLECTIONS.document_embeddings.name, normalizedPoints);
    }

    async searchDocumentEmbeddings(vector, options = {}) {
        return this.search(COLLECTIONS.document_embeddings.name, {
            vector,
            limit: options.limit || 10,
            with_payload: true
        });
    }

    async deleteDocumentEmbeddings(docId) {
        if (Array.isArray(docId)) {
            return this.deleteByIds(COLLECTIONS.document_embeddings.name, docId);
        }
        return this.deleteByDocId(COLLECTIONS.document_embeddings.name, docId);
    }

    // Specific payload update wrapper
    async updatePayloadForDoc(collection, docId, payload) {
        // This wraps the generic updatePayload but targets a specific collection if needed,
        // or we can just use the generic one if the caller passes the collection name.
        // The generic updatePayload iterates all collections, which might be overkill if we know the target.
        // Implementing specific single-collection update:
        try {
            const points = await this._getPointsByDocId(collection, docId);
            if (points.length === 0) return { status: 'skipped', reason: 'no_points' };
            
            const pointIds = points.map(p => p.id);
            const result = await this.circuitBreaker.execute(async () => {
                await this.client.setPayload(collection, {
                    points: pointIds,
                    payload
                });
            });

            if (result.fallback || !result.success) {
                throw result.error || new Error(`Payload update failed for ${collection}`);
            }
            return { status: 'ok', count: pointIds.length };
        } catch (e) {
            logger.error(`[QdrantAdapter] updatePayloadForDoc failed: ${e.message}`);
            throw e;
        }
    }

    // Legacy alias for getPoint (used in tests)
    async getPoint(collection, id) {
        const result = await this.circuitBreaker.execute(async () => {
            // Support both client.getPoint (legacy stub/tests) and client.retrieve (qdrant client's newer name)
            if (typeof this.client.getPoint === 'function') {
                return await this.client.getPoint(collection, id);
            }
            if (typeof this.client.retrieve === 'function') {
                return await this.client.retrieve(collection, {
                    ids: [id],
                    with_payload: true,
                    with_vector: true
                });
            }
            throw new Error('Qdrant client does not support getPoint or retrieve');
        });
        
        if (result.fallback || !result.success) {
            throw result.error || new Error(`getPoint failed for ${collection} id=${id}`);
        }

        // If the underlying client returned an array (retrieve), return first element
        const data = result.data;
        if (Array.isArray(data)) return data[0];
        return data;
    }
}

// Export class, singleton instance, and constants
const qdrantAdapter = new QdrantAdapter();

module.exports = {
    QdrantAdapter,
    qdrantAdapter,
    COLLECTIONS
};

File: containers/text-rag/qdrant_adapter.py
import logging
import os
from typing import Any, Dict, List, Optional, Union

from qdrant_client import QdrantClient
from qdrant_client.models import (
    Distance,
    FieldCondition,
    Filter,
    MatchValue,
    PayloadSchemaType,
    PointStruct,
    VectorParams,
)

logger = logging.getLogger(__name__)


class QdrantAdapter:
    """Adapter for Qdrant vector database operations."""

    def __init__(
        self,
        host: Optional[str] = None,
        port: Optional[int] = None,
        api_key: Optional[str] = None
    ) -> None:
        """Initialize Qdrant client and ensure collection exists."""
        self.host = host or os.getenv("QDRANT_HOST", "qdrant")
        self.port = port or int(os.getenv("QDRANT_PORT", "6333"))
        self.api_key = api_key or os.getenv("QDRANT_API_KEY")

        self.client = QdrantClient(
            host=self.host,
            port=self.port,
            api_key=self.api_key
        )
        self.collection_name = 'document_embeddings'
        self._ensure_collection()

    def _ensure_collection(self) -> None:
        """Ensure collection exists with correct distance metric."""
        try:
            collections = self.client.get_collections()
            exists = any(
                c.name == self.collection_name
                for c in collections.collections
            )

            if exists:
                info = self.client.get_collection(self.collection_name)
                vectors_config = info.config.params.vectors

                # Validate distance metric lock
                current_dist = None
                if isinstance(vectors_config, VectorParams):
                    current_dist = vectors_config.distance
                elif isinstance(vectors_config, dict):
                    current_dist = vectors_config.get('distance')

                # Check against expected COSINE
                if current_dist != Distance.COSINE:
                    msg = (
                        f"Distance Metric Lock violation: Collection "
                        f"'{self.collection_name}' has distance "
                        f"'{current_dist}' but expected 'Cosine'."
                    )
                    logger.error(msg)
                    raise ValueError(msg)

                logger.info(
                    f"Collection '{self.collection_name}' verified."
                )
            else:
                logger.info(
                    f"Creating collection '{self.collection_name}'"
                )
                self.client.create_collection(
                    collection_name=self.collection_name,
                    vectors_config=VectorParams(
                        size=384,
                        distance=Distance.COSINE
                    ),
                )

            self._create_payload_indexes()

        except Exception as e:
            logger.error(f"Failed to ensure collection: {e}")
            raise

    def _create_payload_indexes(self) -> None:
        """Create payload indexes for filtering fields."""
        fields = [
            ("doc_id", PayloadSchemaType.INTEGER),
            ("correspondent_id", PayloadSchemaType.INTEGER),
            ("tag_ids", PayloadSchemaType.INTEGER),
        ]

        for field_name, schema_type in fields:
            try:
                self.client.create_payload_index(
                    collection_name=self.collection_name,
                    field_name=field_name,
                    field_schema=schema_type,
                )
                logger.info(f"Ensured index for field '{field_name}'")
            except Exception as e:
                # Log warning but continue if index creation fails
                # (e.g. already exists)
                logger.warning(
                    f"Could not create index for '{field_name}': {e}"
                )

    def upsert_document(
        self,
        doc_id: int,
        embedding: Union[List[float], Any],
        metadata: Dict[str, Any]
    ) -> None:
        """Upsert a document embedding with metadata."""
        try:
            # Ensure embedding is a list (handle numpy arrays)
            if hasattr(embedding, 'tolist'):
                embedding = embedding.tolist()

            payload = {
                'doc_id': doc_id,
                'title': metadata.get('title'),
                'correspondent_id': metadata.get('correspondent_id'),
                'tag_ids': metadata.get('tag_ids', []),
                'created': metadata.get('created'),
                'content_preview': metadata.get('content', '')[:200]
            }

            point = PointStruct(
                id=doc_id,
                vector=embedding,
                payload=payload
            )

            self.client.upsert(
                collection_name=self.collection_name,
                points=[point]
            )
            logger.info(f"Upserted document {doc_id} to Qdrant.")
        except Exception as e:
            logger.error(f"Failed to upsert document {doc_id}: {e}")
            raise

    def upsert_documents(
        self,
        documents: List[Dict[str, Any]]
    ) -> None:
        """Batch upsert documents."""
        try:
            points = []
            for doc in documents:
                doc_id = doc['doc_id']
                embedding = doc['embedding']
                metadata = doc['metadata']
                
                if hasattr(embedding, 'tolist'):
                    embedding = embedding.tolist()

                payload = {
                    'doc_id': doc_id,
                    'title': metadata.get('title'),
                    'correspondent_id': metadata.get('correspondent_id'),
                    'tag_ids': metadata.get('tag_ids', []),
                    'created': metadata.get('created'),
                    'content_preview': metadata.get('content', '')[:200]
                }

                points.append(PointStruct(
                    id=doc_id,
                    vector=embedding,
                    payload=payload
                ))

            if points:
                self.client.upsert(
                    collection_name=self.collection_name,
                    points=points
                )
                logger.info(f"Upserted batch of {len(points)} documents to Qdrant.")
        except Exception as e:
            logger.error(f"Failed to batch upsert documents: {e}")
            raise

    def search(
        self,
        query_embedding: Union[List[float], Any],
        k: int = 5,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """Search for similar documents with optional filters."""
        try:
            # Ensure query_embedding is a list
            if hasattr(query_embedding, 'tolist'):
                query_embedding = query_embedding.tolist()

            query_filter = None
            if filters:
                conditions = []
                if 'correspondent_id' in filters:
                    conditions.append(
                        FieldCondition(
                            key='correspondent_id',
                            match=MatchValue(
                                value=filters['correspondent_id']
                            )
                        )
                    )

                if 'tag_ids' in filters and filters['tag_ids']:
                    # AND logic: document must have all provided tags
                    for tag_id in filters['tag_ids']:
                        conditions.append(
                            FieldCondition(
                                key='tag_ids',
                                match=MatchValue(value=tag_id)
                            )
                        )

                if conditions:
                    query_filter = Filter(must=conditions)

            results = self.client.query_points(
                collection_name=self.collection_name,
                query=query_embedding,
                query_filter=query_filter,
                limit=k
            ).points

            return [
                {
                    'doc_id': point.id,
                    'score': point.score,
                    'payload': point.payload
                }
                for point in results
            ]
        except Exception as e:
            logger.error(f"Search failed: {e}")
            raise

    def update_payload(
        self,
        doc_id: int,
        metadata: Dict[str, Any]
    ) -> None:
        """Update payload for a specific document."""
        try:
            update_data = {}
            if 'correspondent_id' in metadata:
                update_data['correspondent_id'] = metadata['correspondent_id']
            if 'tag_ids' in metadata:
                update_data['tag_ids'] = metadata['tag_ids']

            if not update_data:
                return

            self.client.set_payload(
                collection_name=self.collection_name,
                payload=update_data,
                points=[doc_id]
            )
            logger.info(f"Updated payload for document {doc_id}.")
        except Exception as e:
            logger.error(
                f"Failed to update payload for document {doc_id}: {e}"
            )
            raise

    def get_collection_stats(self) -> Dict[str, Any]:
        """Get statistics for the document_embeddings collection."""
        try:
            info = self.client.get_collection(self.collection_name)
            return {
                "exists": True,
                "point_count": info.points_count,
                "status": info.status
            }
        except Exception:
            return {"exists": False, "point_count": 0}

    def health_check(self) -> Dict[str, Any]:
        """Check Qdrant connectivity and collection status.

        Returns a dict with the overall `healthy` boolean, a mapping of
        `collections` to their status, and an aggregated `point_count`.
        """
        try:
            collections = self.client.get_collections()
            collection_names = [c.name for c in collections.collections]

            details: Dict[str, Any] = {}
            total_points = 0

            # Gather stats for all collections reported by the server
            for cname in collection_names:
                try:
                    info = self.client.get_collection(cname)
                    points = int(getattr(info, "points_count", 0) or 0)
                    details[cname] = {
                        "exists": True,
                        "point_count": points,
                        "status": str(info.status),
                    }
                    total_points += points
                except Exception as e:
                    # If a single collection query fails, still continue
                    logger.warning(
                        "Could not fetch stats for collection '%s': %s",
                        cname,
                        str(e),
                    )
                    details[cname] = {
                        "exists": False,
                        "point_count": 0,
                        "status": "error",
                    }

            # Ensure our expected collection is present in the response
            if self.collection_name not in details:
                try:
                    info = self.client.get_collection(self.collection_name)
                    points = int(getattr(info, "points_count", 0) or 0)
                    details[self.collection_name] = {
                        "exists": True,
                        "point_count": points,
                        "status": str(info.status),
                    }
                    total_points += points
                except Exception:
                    details[self.collection_name] = {
                        "exists": False,
                        "point_count": 0,
                        "status": "missing",
                    }

            return {
                "healthy": True,
                "collections": details,
                "point_count": total_points,
            }
        except Exception as e:
            logger.error("Health check failed: %s", str(e))
            return {"healthy": False, "error": str(e)}


# Singleton instance
qdrant_adapter = QdrantAdapter()
